File: packages/sdk-ts/src/viz/matrix.ts
/**
 * Matrix views — standalone functions operating on ManifestInstance.
 *
 * Extracted from ManifestInstance to keep the kernel class focused.
 */

import type { ManifestInstance } from "../kernel/instance.js";

// ---------------------------------------------------------------------------
// Matrix (structured)
// ---------------------------------------------------------------------------

export function matrix(mi: ManifestInstance): Record<string, unknown> {
  const tree = mi.dependencyTree();
  if (Object.keys(tree).length === 0) {
    return { agents: [], dependencies: {}, matrix: {} };
  }

  const agents = Object.keys(tree).sort();

  // Collect all deps grouped by kind
  const depsByKind = new Map<string, Set<string>>();
  for (const info of Object.values(tree)) {
    const treeInfo = info as Record<string, unknown>;
    const dependsOn = treeInfo.depends_on as Record<string, Record<string, Record<string, unknown>>> | undefined;
    if (!dependsOn) continue;
    for (const deps of Object.values(dependsOn)) {
      for (const [depName, depInfo] of Object.entries(deps)) {
        const kind = (depInfo.kind as string) ?? "Unknown";
        if (!depsByKind.has(kind)) depsByKind.set(kind, new Set());
        depsByKind.get(kind)!.add(depName);
      }
    }
  }

  // Build matrix
  const matrixData: Record<string, Record<string, boolean>> = {};
  for (const [agentName, info] of Object.entries(tree)) {
    const treeInfo = info as Record<string, unknown>;
    const dependsOn = treeInfo.depends_on as Record<string, Record<string, unknown>> | undefined;
    const row: Record<string, boolean> = {};
    if (dependsOn) {
      for (const deps of Object.values(dependsOn)) {
        for (const depName of Object.keys(deps)) {
          row[depName] = true;
        }
      }
    }
    matrixData[agentName] = row;
  }

  // Build sorted dependencies object
  const dependencies: Record<string, string[]> = {};
  const sortedKinds = [...depsByKind.keys()].sort();
  for (const kind of sortedKinds) {
    dependencies[kind] = [...depsByKind.get(kind)!].sort();
  }

  return { agents, dependencies, matrix: matrixData };
}

// ---------------------------------------------------------------------------
// Matrix Markdown
// ---------------------------------------------------------------------------

export function matrixMarkdown(mi: ManifestInstance): string {
  const data = matrix(mi);
  const agents = data.agents as string[];
  if (agents.length === 0) {
    return "No agents with dependencies found.";
  }

  const dependencies = data.dependencies as Record<string, string[]>;
  const matrixData = data.matrix as Record<string, Record<string, boolean>>;

  // Flatten all dep names in kind order
  const allDeps: Array<[string, string]> = []; // [kind, name]
  for (const [kind, names] of Object.entries(dependencies)) {
    for (const name of names) {
      allDeps.push([kind, name]);
    }
  }

  // Header
  const depHeaders = allDeps.map(([, n]) => n);
  const header = `| Agent | ${depHeaders.join(" | ")} |`;
  const sep = `|-------|${allDeps.map(() => ":---:").join("|")}|`;

  const rows: string[] = [];
  for (const agent of agents) {
    const cells = allDeps.map(([, depName]) =>
      matrixData[agent]?.[depName] ? "\u25CF" : " "
    );
    rows.push(`| **${agent}** | ${cells.join(" | ")} |`);
  }

  // Kind legend row
  const kindRow = `| *Kind* | ${allDeps.map(([k]) => `*${k[0].toUpperCase()}*`).join(" | ")} |`;

  return [header, sep, kindRow, ...rows].join("\n");
}

File: packages/sdk-py/dna/viz/matrix.py
"""Matrix views — standalone functions operating on ManifestInstance.

Extracted from ManifestInstance to keep the kernel class focused.
1:1 parity with TypeScript viz/matrix.ts.
"""
from __future__ import annotations
from typing import Any, TYPE_CHECKING

if TYPE_CHECKING:
    from dna.kernel.instance import ManifestInstance


def matrix(mi: ManifestInstance) -> dict[str, Any]:
    """Agent x dependency matrix showing which agent uses what.

    Returns:
        {
            "agents": ["swe-agent", "reviewer-agent"],
            "dependencies": {"Skill": ["pr-review", ...], "Guardrail": [...], ...},
            "matrix": {
                "swe-agent": {"pr-review": true, "safety": true, ...},
                "reviewer-agent": {"pr-review": true, ...}
            }
        }
    """
    tree = mi.dependency_tree()
    if not tree:
        return {"agents": [], "dependencies": {}, "matrix": {}}

    agents = sorted(tree.keys())
    # Collect all deps grouped by kind
    deps_by_kind: dict[str, set[str]] = {}
    for info in tree.values():
        for deps in info.get("depends_on", {}).values():
            for dep_name, dep_info in deps.items():
                kind = dep_info.get("kind", "Unknown")
                deps_by_kind.setdefault(kind, set()).add(dep_name)

    # Build matrix
    matrix_data: dict[str, dict[str, bool]] = {}
    for agent_name, info in tree.items():
        row: dict[str, bool] = {}
        for deps in info.get("depends_on", {}).values():
            for dep_name in deps:
                row[dep_name] = True
        matrix_data[agent_name] = row

    return {
        "agents": agents,
        "dependencies": {k: sorted(v) for k, v in sorted(deps_by_kind.items())},
        "matrix": matrix_data,
    }


def matrix_markdown(mi: ManifestInstance) -> str:
    """Agent x dependency matrix as a Markdown table."""
    data = matrix(mi)
    if not data["agents"]:
        return "No agents with dependencies found."

    # Flatten all dep names in kind order
    all_deps: list[tuple[str, str]] = []  # (kind, name)
    for kind, names in data["dependencies"].items():
        for name in names:
            all_deps.append((kind, name))

    # Header
    dep_headers = [f"{n}" for _, n in all_deps]
    header = "| Agent | " + " | ".join(dep_headers) + " |"
    sep = "|-------|" + "|".join([":---:" for _ in all_deps]) + "|"

    rows = []
    for agent in data["agents"]:
        cells = []
        for _, dep_name in all_deps:
            cells.append("\u25cf" if data["matrix"].get(agent, {}).get(dep_name) else " ")
        rows.append(f"| **{agent}** | " + " | ".join(cells) + " |")

    # Kind legend row
    kind_row = "| *Kind* | " + " | ".join(
        f"*{k[0].upper()}*" for k, _ in all_deps
    ) + " |"

    return "\n".join([header, sep, kind_row, *rows])
